### PR TITLE
ISPN-11176 Xsite Max Idle

### PR DESCRIPTION
**Pull request by wburns**
https://issues.redhat.com/browse/ISPN-11176
https://issues.redhat.com/browse/ISPN-12337

### CODE REVIEW

**Comment by wburns** [comment]
Added a new commit to refactor touch to be a visitable command which provides handling for an xsite invoked touch command to be forwarded to the appropriate owner as well as topology change updates.

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java [hunk 1]
```diff
@@ -257,6 +260,25 @@ private XSiteStatePushCommand newStatePushCommand(List<XSiteState> stateList) {
       return defaultHandler.cache().clearAsync();
    }
 
+   @Override
+   public CompletionStage<Boolean> touchEntry(Object key) {
+      int segment = clusteringDependentLogic.getCacheTopology().getSegment(key);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:265]: shouldn't be sent to the primary owner?

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java [hunk 2]
```diff
@@ -126,9 +130,8 @@ public ClusteredCacheBackupReceiver(String cacheName) {
    public void start() {
       //it would be nice if we could inject bootstrap component
       //this feels kind hacky but saves 3 fields in this class
-      ComponentRegistry cr = cache.getAdvancedCache().getComponentRegistry();
-      TransactionHandler txHandler = new TransactionHandler(cache, cr.getTransactionTable());
-      defaultHandler = new DefaultHandler(txHandler, cr.getComponent(BlockingManager.class));
+      TransactionHandler txHandler = new TransactionHandler(cache, componentRegistry.getTransactionTable());
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:133]: can you revert this and remove the `componentRegistry' field? I know I'm complaining just to save 8 bytes eh eh

**Review by pruivo — CHANGES_REQUESTED** [review]
first batch :) didn't check everything yet

File: core/src/main/java/org/infinispan/cache/impl/CacheImpl.java [hunk 3]
```diff
@@ -1218,6 +1220,24 @@ public String toString() {
...
+      return touch(key, -1, touchEvenIfExpired, 0);
+   }
+
+   @Override
+   public CompletionStage<Boolean> touch(Object key, int segment, boolean touchEvenIfExpired) {
+      return touch(key, segment, touchEvenIfExpired, 0);
+   }
+
+   public CompletionStage<Boolean> touch(Object key, int segment, boolean touchEvenIfExpired, long flagBitSet) {
+      if (segment == -1) {
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/CacheImpl.java:1234]: Sure

File: core/src/main/java/org/infinispan/cache/impl/CacheImpl.java [hunk 4]
```diff
@@ -1218,6 +1220,24 @@ public String toString() {
       return "Cache '" + name + "'@" + (config != null && config.clustering().cacheMode().isClustered() ? getRpcManager().getAddress() : Util.hexIdHashCode(getCacheManager()));
    }
 
+   @Override
+   public CompletionStage<Boolean> touch(Object key, boolean touchEvenIfExpired) {
+      return touch(key, -1, touchEvenIfExpired, 0);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/CacheImpl.java:1225]: 👍

File: core/src/main/java/org/infinispan/cache/impl/CacheImpl.java [hunk 5]
```diff
@@ -1218,6 +1220,24 @@ public String toString() {
...
+   }
+
+   @Override
+   public CompletionStage<Boolean> touch(Object key, int segment, boolean touchEvenIfExpired) {
+      return touch(key, segment, touchEvenIfExpired, 0);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/CacheImpl.java:1230]: 👍

File: core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java [hunk 6]
```diff
@@ -390,6 +392,26 @@ public void setAvailability(AvailabilityMode availabilityMode) {
       throw new UnsupportedOperationException();
    }
 
+   @Override
+   public CompletionStage<Boolean> touch(Object key, boolean touchEvenIfExpired) {
+      Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED);
+      int segment = keyPartitioner.getSegment(key);
+      return touch(key, segment, touchEvenIfExpired);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:399]: I was just being consistent with the rest of the `SimpleCacheImpl` methods.

File: core/src/main/java/org/infinispan/commands/CommandsFactoryImpl.java [hunk 7]
```diff
@@ -694,8 +695,8 @@ public CheckTransactionRpcCommand buildCheckTransactionRpcCommand(Collection<Glo
    }
 
    @Override
-   public TouchCommand buildTouchCommand(Object key, int segment) {
-      return new TouchCommand(cacheName, key, segment);
+   public TouchCommand buildTouchCommand(Object key, int segment, boolean touchEvenIfExpired, long flagBitSet) {
+      return new TouchCommand(key, segment, flagBitSet, touchEvenIfExpired);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/commands/CommandsFactoryImpl.java:699]: Not currently, the idea was we may at some point want a LOCAL forced touch for example.

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java (hunk 2 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:133]: Okay.

File: core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java (hunk 1 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/ClusteredCacheBackupReceiver.java:265]: Hrmm, yeah this should invoke the touch command. Let me do that.

File: core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:399]: But now that I look again, I agree it isn't required.

File: core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java [hunk 8]
```diff
@@ -390,6 +392,26 @@ public void setAvailability(AvailabilityMode availabilityMode) {
...
+   }
+
+   @Override
+   public CompletionStage<Boolean> touch(Object key, int segment, boolean touchEvenIfExpired) {
+      Objects.requireNonNull(key, NULL_KEYS_NOT_SUPPORTED);
+      InternalCacheEntry<K, V> entry = dataContainer.peek(segment, key);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/cache/impl/SimpleCacheImpl.java:405]: I added a `segment < 0` check.

File: core/src/main/java/org/infinispan/commands/CommandsFactoryImpl.java (hunk 7 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/commands/CommandsFactoryImpl.java:699]: yeah, I was just worried if the user invokes the `touch()` with some weird flags (like, skip remote)

File: core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+      eventuallyEquals("v", () -> cache(LON, 1).get(key));
+      Cache<Object, Object> backupCache = backup(LON);
+      assertNull(backupCache.get("k"));
+      eventuallyEquals("v", () -> backup(LON).get(key));
+
+      // Now expire the entry
+      timeService.advance(TimeUnit.SECONDS.toMillis(2));
+      assertNull(cache.get(key));
+
+      // TODO: check for the touch command being invoked on the remote site for lifespan = false
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java:171]: actually, I would inspect the DataContainer to ensure the InternalCacheEntry's lastUsed is the expected value

File: core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java [hunk 10]
```diff
@@ -30,32 +37,159 @@
...
+      doTest(method, new TestOperationInterop(TestOperation.REMOVE));
    }
 
    public void testConditionalRemove(Method method) {
-      doTest(method, TestOperation.REMOVE_CONDITIONAL);
+      doTest(method, new TestOperationInterop(TestOperation.REMOVE_CONDITIONAL));
+   }
+
+   // TODO: need to do this still?
+   //   public void testMaxIdleExpirationASync(Method method) {
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java:102]: what does sync/async mean here? the xsite is async only in this class.

**Review by pruivo — CHANGES_REQUESTED** [review]
I still have a couple of classes to check but it is late and my eyes are tired. :) 
I'll try to finish it tomorrow.

File: core/src/main/java/org/infinispan/container/impl/DefaultDataContainer.java [hunk 11]
```diff
@@ -233,4 +232,10 @@ public void cleanUp() {
          evictionCache.cleanUp();
       }
    }
+
+
+   @Override
+   public void forEachSegment(ObjIntConsumer<PeekableTouchableMap<K, V>> segmentMapConsumer) {
+      segmentMapConsumer.accept(entries, 0);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/DefaultDataContainer.java:239]: The point is to iterate over all the various segmented maps of the container. I can't see having the API not have the segment.

And the reason for 0 is because this container has a single map. This is inline with what the `SingleSegmentKeyPartitioner` does as well.

File: core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java [hunk 12]
```diff
@@ -84,7 +85,7 @@ public void init() {
             }
          } else if (isOwner || readEntry.isL1Entry()) {
             if (readEntry.canExpire()) {
-               CompletionStage<Boolean> expiredStage = expirationManager.handlePossibleExpiration(readEntry, segment, false);
+               CompletionStage<Boolean> expiredStage = expirationManager.handlePossibleExpiration(readEntry, segment, hasLock);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java:88]: It doesn't skip the max-idle expiration check. What it does is change the flags on the cache that the remove (max idle | lifespan) expired command is invoked on. For pessimistic transactions if it already has the lock we can't fire the expiration command and get a lock as it is already held.

https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/expiration/impl/TxClusterExpirationManager.java#L37

File: core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java [hunk 13]
```diff
@@ -107,4 +108,9 @@ public int sizeIncludingExpired() {
    public void clear() {
       map.clear();
    }
+
+   @Override
+   public void forEachSegment(ObjIntConsumer<PeekableTouchableMap<WrappedBytes, WrappedBytes>> segmentMapConsumer) {
+      segmentMapConsumer.accept(map, 0);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/offheap/OffHeapDataContainer.java:114]: I think 0 is the correct value as this is what we do for other cases when we don't have segmentation.

File: core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java [hunk 14]
```diff
@@ -99,6 +88,9 @@ public void start() {
                   expWakeUpInt, expWakeUpInt, TimeUnit.MILLISECONDS);
          }
       }
+      // Data container entries are retrieved directly, so we don't need to worry about an encodings
+      this.cache = AbstractDelegatingCache.unwrapCache(cacheRef.wired()).getAdvancedCache();
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java:92]: It used to, believe it still does. This was just moved over from the ClusterExpirationManager as they both use the cache instance now.

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java [hunk 15]
```diff
@@ -592,8 +593,11 @@ public Object visitRemoveExpiredCommand(InvocationContext ctx, RemoveExpiredComm
          } else if (trace) {
             log.trace("Cannot remove entry as its lifespan or value do not match");
          }
-      } else if (trace) {
-         log.trace("Nothing to remove since the entry doesn't exist in the context or it is already removed");
+      } else {
+         if (trace) {
+            log.trace("Nothing to remove since the entry doesn't exist in the context or it is already removed - assume command was successful");
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:598]: We don't want to let something like an eviction to cause the expiration command to fail. If the value isn't present we have to assume it can be expired.

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java [hunk 16]
```diff
@@ -1115,6 +1119,32 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteM
...
+   @Override
+   public Object visitTouchCommand(InvocationContext ctx, TouchCommand command) throws Throwable {
+      int segment = command.getSegment();
+      Object key = command.getKey();
+      InternalCacheEntry<?, ?> ice = dataContainer.peek(segment, key);
+      if (ice == null) {
+         if (trace) {
+            log.tracef("Entry was not in the container to touch for key %s", key);
+         }
+         return Boolean.FALSE;
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:1131]: It will be boxed anyways, may as well be explicit :)

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java [hunk 17]
```diff
@@ -1115,6 +1119,32 @@ public Object visitReadWriteManyEntriesCommand(InvocationContext ctx, ReadWriteM
       return returns;
    }
 
+   @Override
+   public Object visitTouchCommand(InvocationContext ctx, TouchCommand command) throws Throwable {
+      int segment = command.getSegment();
+      Object key = command.getKey();
+      InternalCacheEntry<?, ?> ice = dataContainer.peek(segment, key);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:1126]: No, unfortunately. We don't want to touch in the case of `isTouchEvenIfExpired` is false. We could have different code paths where if this is true we just call peek if you really wanted though.

File: core/src/main/java/org/infinispan/interceptors/xsite/BaseBackupInterceptor.java [hunk 18]
```diff
@@ -51,6 +56,38 @@ Object invokeNextAndWaitForCrossSite(TxInvocationContext<?> ctx, VisitableComman
...
+      int segment = command.getSegment();
+      DistributionInfo dInfo = clusteringDependentLogic.getCacheTopology().getSegmentDistribution(segment);
+      // Only require primary to check remote site and add to irac queue - If primary dies then a backup will end up
+      // doing the same as promoted primary - We also don't add tracked up to backup as we don't care if the
+      // remove expired is lost due to topology change it just will cause another check later but maintain consistency
+      if (dInfo.isPrimary()) {
+         CompletionStage<Boolean> expired = iracManager.checkAndTrackExpiration(command.getKey());
+         return asyncValue(expired).thenApply(ctx, command, (rCtx, rCommand, rv) -> {
+            if ((Boolean) rv) {
+               return invokeNextThenAccept(rCtx, command, this::handleExpiredReturn);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/xsite/BaseBackupInterceptor.java:76]: Did you want me to make the lambda an object as well? :D I personally don't think it is worth it as this shouldn't happen that often.

[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/xsite/BaseBackupInterceptor.java:76]: Oh you mention that below :)

File: core/src/main/java/org/infinispan/interceptors/xsite/BaseBackupInterceptor.java [hunk 19]
```diff
@@ -51,6 +56,38 @@ Object invokeNextAndWaitForCrossSite(TxInvocationContext<?> ctx, VisitableComman
...
+      // a read that we aren't aware of - so we must synchronously remove the entry if expired on the remote site
+      // and if it isn't expired on the remote site we must update the access time locally here
+      int segment = command.getSegment();
+      DistributionInfo dInfo = clusteringDependentLogic.getCacheTopology().getSegmentDistribution(segment);
+      // Only require primary to check remote site and add to irac queue - If primary dies then a backup will end up
+      // doing the same as promoted primary - We also don't add tracked up to backup as we don't care if the
+      // remove expired is lost due to topology change it just will cause another check later but maintain consistency
+      if (dInfo.isPrimary()) {
+         CompletionStage<Boolean> expired = iracManager.checkAndTrackExpiration(command.getKey());
+         return asyncValue(expired).thenApply(ctx, command, (rCtx, rCommand, rv) -> {
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/xsite/BaseBackupInterceptor.java:74]: I personally don't think it is worth it given how infrequently it should be called, but I can add it in.

File: core/src/main/java/org/infinispan/xsite/status/DefaultTakeOfflineManager.java [hunk 20]
```diff
@@ -197,6 +204,10 @@ public void siteOnline() {
       @Override
       public void siteOffline() {
          getEventLogger().info(EventLogCategory.CLUSTER, MESSAGES.siteOffline(siteName));
+         log.debug("Touching all in memory entries as a site has gone offline");
+         long currentTimeMillis = timeService.wallClockTime();
+         dataContainer.forEachSegment((map, segment) ->
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/xsite/status/DefaultTakeOfflineManager.java:209]: Hrmm, yeah we can add that to the `PeekableTouchableMap` interface :)

File: core/src/test/java/org/infinispan/functional/decorators/FunctionalAdvancedCache.java [hunk 21]
```diff
@@ -282,6 +282,16 @@ public void start() {
...
    }
 
+   @Override
+   public CompletionStage<Boolean> touch(Object key, int segment, boolean touchEvenIfExpired) {
+      return cache.touch(key, segment, touchEvenIfExpired);
+   }
+
+   @Override
+   public CompletionStage<Boolean> touch(Object key, boolean touchEvenIfExpired) {
+      return null;
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/functional/decorators/FunctionalAdvancedCache.java:292]: I changed it.

File: core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java [hunk 22]
```diff
@@ -0,0 +1,260 @@
...
+      MagicKey key = readOnPrimary ? new MagicKey(cache) : new MagicKey(cache(LON, 1));
+      if (lifespan) {
+         cache.put(key, "v", 1, TimeUnit.SECONDS);
+      } else {
+         cache.put(key, "v", -1, TimeUnit.SECONDS, 1, TimeUnit.SECONDS);
+      }
+      eventuallyEquals("v", () -> cache(LON, 0).get(key));
+      eventuallyEquals("v", () -> cache(LON, 1).get(key));
+      Cache<Object, Object> backupCache = backup(LON);
+      assertNull(backupCache.get("k"));
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java:164]: Refactoring mistake, thought I had removed it.

File: core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java [hunk 23]
```diff
@@ -0,0 +1,260 @@
...
+      eventuallyEquals("v", () -> backup(LON).get(key));
+
+      // Now expire the entry
+      timeService.advance(TimeUnit.SECONDS.toMillis(2));
+      assertNull(cache.get(key));
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java:169]: Sure

File: core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java (hunk 10 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java:102]: I meant the remove expired command itself. The test waits for its completion. If it didn't then we could have a remove expired and other puts running concurrently.

File: core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java (hunk 9 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java:171]: Yeah, that is what I meant by it actually.

Comment (wburns) [reply core/src/test/java/org/infinispan/xsite/AsyncBackupExpirationTest.java:171]: I moved it up a bit before the expiration to verify the last access was correct.

File: core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java [hunk 24]
```diff
@@ -194,7 +189,13 @@ private void visitKey(Object key, WriteCommand command) {
       if (getOwnership(segment) != Ownership.PRIMARY) {
          return;
       }
-      IracMetadata metadata = iracVersionGenerator.generateNewMetadata(segment);
+      IracMetadata metadata;
+      // RemoveExpired should lose to any other conflicting write
+      if (command instanceof RemoveExpiredCommand) {
+         metadata = iracVersionGenerator.generateMetadataWithCurrentVersion(segment);
```
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java]: I have done a mixture, because unfortunately with eviction it is possible the entry is not in the container as well.

**Comment by wburns** [comment]
I have pushed a commit that needs to be rebased that addresses the most recent comments. Thanks again @pruivo !

File: core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java:88]: I'm looking at https://github.com/infinispan/infinispan/blob/master/core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java#L261
in pessimistic cache, `checkExpiredMaxIdle(ice, segment)` isn't invoked because `isWrite==true`

[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java:88]: That is because the entry wasn't expired. In this case we don't want to touch the entry on a write as it will be overwritten anyways.

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:1126]: `isTouchEvenIfExpired` could be a parameter :)

[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:1126]: Hrmm? It is in the `TouchCommand`.

File: core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java [hunk 25]
```diff
@@ -70,21 +63,16 @@
...
    private static final int MAX_CONCURRENT_EXPIRATIONS = 100;
 
-   @Inject protected ComponentRef<AdvancedCache<K, V>> cacheRef;
    @Inject protected RpcManager rpcManager;
    @Inject protected DistributionManager distributionManager;
 
-   protected AdvancedCache<K, V> cache;
    private Address localAddress;
    private long timeout;
    private String cacheName;
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java:71]: can be removed. it is available in super class

File: core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java [hunk 26]
```diff
@@ -34,6 +48,120 @@
...
+         List<Address> owners = isScattered ? cacheTopology.getActualMembers() : info.readOwners();
+         AbstractTouchResponseCollector collector = isScattered ? ScatteredTouchResponseCollector.INSTANCE :
+               TouchResponseCollector.INSTANCE;
+         CompletionStage<Boolean> remoteInvocation = rpcManager.invokeCommand(owners, command, collector,
+               rpcManager.getSyncRpcOptions());
+         return invokeNextThenApply(ctx, command, (rCtx, rCommand, rValue) -> {
+            Boolean touchedLocally = (Boolean) rValue;
+            Boolean collectedResponse = collector.addBooleanResponse(null, touchedLocally);
+            if (collectedResponse != null) {
+               return asyncValue(remoteInvocation.thenApply(ignore -> touchedLocally));
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java:151]: why do you need to wait for the remote operation? can't you return `touchedLocally` directly?

File: core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java [hunk 27]
```diff
@@ -189,12 +186,26 @@ private boolean skipCommand(InvocationContext ctx, FlagAffectedCommand command)
...
+      // RemoveExpired should lose to any other conflicting write
+      if (command instanceof RemoveExpiredCommand) {
+         CacheEntry<?, ?> ce = ctx.lookupEntry(key);
+         PrivateMetadata pm = ce.getInternalMetadata();
+         if (pm != null) {
+            metadata = pm.iracMetadata();
+         }
+
+         if (metadata == null) {
+            metadata = iracVersionGenerator.generateMetadataWithCurrentVersion(segment);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java:204]: I'm going to take some time to think about this... not 100% convinced eh eh
also, you could check `iracVersionGenerator.getTombstone(key);` before invoking the "generate" method

**Review by pruivo — COMMENTED** [review]
minor comments

File: core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java (hunk 10 above)
Comment (pruivo) [reply core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java:102]: I'm not sure if it matters. the IracManager is blocking everything until the end. You don't have the remove/put going to the other sites.

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:1126]: I mean `DataContainer.touch(segment, key, currentTime,command.isTouchEvenIfExpired())`

File: core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java:88]: yeah, but a get+force_write_lock doesn't mean the key will be written in the transaction. Not sure if I'm making my concern clear...

File: core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java (hunk 10 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/xsite/irac/Irac3SitesConflictTest.java:102]: It shouldn't block the invoking thread. It returns a CompletionStage.

File: core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java (hunk 26 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/ClusteringInterceptor.java:151]: It needs to be sync to maintain the cluster max idle guarantees that if a node goes down after a touch that it won't expire early.

File: core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java (hunk 27 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java:204]: Sure, I trust you on it ;)

File: core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java:1126]: Hrmm, sure, that could work.

File: core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (pruivo) [reply core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java:88]: I made the following test: https://github.com/pruivo/infinispan/blob/8d5d98492df393c344fe2707dcd12dcf7498464f/core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java

* tx read + rollback, touches the key (it makes senses IMO)
* tx write + rollback, doesn't touch the key (it should IMO)
* tx read on expired key +  rollback, StackOverflowException (bug!) `org.infinispan.expiration.impl.ClusterExpirationManager.entryExpiredInMemory`
* tx write on expired key +  rollback, it is fine!

File: core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java (hunk 27 above)
Comment (pruivo) [reply core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java:204]: I'm writing a test to make sure it is correct... but now I have a question. why do you need to send the remove to the remote site? 
`org.infinispan.xsite.irac.IracManager#checkAndTrackExpiration()` already makes sure the key is expired everywhere. You can avoid sending the remove expired to the remote site and let them expired the key locally when they access it (or by the expiration reaper?). 
Am I missing something?

Comment (wburns) [reply core/src/main/java/org/infinispan/interceptors/impl/NonTxIracLocalSiteInterceptor.java:204]: We technically don't have to, but it seemed better for us just to send a single IRAC remove expired command instead of waiting for the other site to see the value is expired then ask the original site if it is expired synchronously again.

File: core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (wburns) [reply core/src/main/java/org/infinispan/container/impl/EntryFactoryImpl.java:88]: So I am changing my mind on #2. Originally max idle didn't block the user read operation, but now it does so I don't believe it makes any sense any longer.

#3 is because we need to pass the hasLock variable on a read when using Pessimistic tx with WRITE_LOCK flag present. So it isn't just any tx read :)

**Comment by wburns** [comment]
Also it looks like the recent changes per the review has caused https://ci.infinispan.org/job/Infinispan/job/PR-8699/9/testReport/junit/org.infinispan.xsite/AsyncBackupExpirationTest/testExpiredAccess_LON_NON_TX__NYC_OPTIMISTIC_TX_RC___true__false__/ so I will need to see what happened there too.

**Comment by wburns** [comment]
Updated per new review comments.

File: core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java [hunk 28]
```diff
@@ -269,11 +261,11 @@ private void deleteFromStoresAndNotify(K key, V value, Metadata metadata, Privat
...
     * @param entry the entry to check expiration and touch
     * @param segment the segment the entry maps to
+    * @param currentTime the current time in milliseconds
     * @return whether the entry was expired or not
     */
-   protected CompletionStage<Boolean> checkExpiredMaxIdle(InternalCacheEntry entry, int segment) {
-      TouchCommand touchCommand = cf.running().buildTouchCommand(entry.getKey(), segment);
-      CompletionStage<Boolean> future = (CompletionStage) touchCommand.invokeAsync(componentRegistry);
+   protected CompletionStage<Boolean> checkExpiredMaxIdle(InternalCacheEntry entry, int segment, long currentTime) {
+      CompletionStage<Boolean> future = cache.touch(entry.getKey(), segment, true);
```
Comment (pruivo) [inline core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java:268]: shouldn't `touchEvenIfExpired` be  `false` here?

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
CI is failing :shushing_face:

File: core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java [hunk 29]
```diff
@@ -0,0 +1,181 @@
...
+      assertEquals(value, cache.put(key, value2, -1, TimeUnit.SECONDS, MAX_IDLE, TimeUnit.MILLISECONDS));
+      assertEquals(value2, cache.get(key));
+      // rollback the transaction
+      cache.getAdvancedCache().getTransactionManager().rollback();
+
+      lastWallClock = timeService.wallClockTime();
+      assertNotExpired(key);
+      assertLastUsed(key, lastWallClock);
+
+      // if the tx didn't touch the key, this advance should expire it
```
Comment (pruivo) [inline core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java:70]: Actually, the comments are wrong (same comments, in all tests) :) my bad!

Not sure you notice, but there is a test missing: 
https://github.com/pruivo/infinispan/commit/1cdac799994e45fc5c01ae5ff6a5c3bf6346351e#diff-5932593a9461e5deec52370525a2cf35R144

File: core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java (hunk 28 above)
Comment (wburns) [reply core/src/main/java/org/infinispan/expiration/impl/ExpirationManagerImpl.java:268]: No, it is to cover the edge case of the entry expiring between when we check and then actually touching. We just don't want to touch for things like the xsite max idle check.

File: core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java (hunk 29 above)
Comment (wburns) [reply core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java:70]: > Not sure you notice, but there is a test missing:
> [pruivo@1cdac79#diff-5932593a9461e5deec52370525a2cf35R144](https://github.com/pruivo/infinispan/commit/1cdac799994e45fc5c01ae5ff6a5c3bf6346351e#diff-5932593a9461e5deec52370525a2cf35R144)

Weird, I copied the file directly, must have had two commits before?

Comment (pruivo) [reply core/src/test/java/org/infinispan/expiration/impl/MaxIdlePessimisticTxTest.java:70]:
```suggestion
      // the tx touches the key, this advance should not expire it
```

**Comment by wburns** [comment]
> CI is failing 🤫

Looks like the failure is because the hibernate test assumes the future is completed immediately. But they were trying to write to an expired key so it had to wait until that completed first :)

**Comment by wburns** [comment]
Updated.

**Review by pruivo — APPROVED** [review]

**Comment by pruivo** [comment]
@wburns it looks good to me. Let's wait for CI. :tada:

**Comment by pruivo** [comment]
@wburns 2 failures:
CacheExpiryTest
AsyncBackupExpirationTest

**Comment by wburns** [comment]
I can't reproduce the AsyncBackupExpirationTest for the life of me. Interestingly the part that failed was when using lifespan, which shouldn't be an issue. I will keep trying though.

And CacheExpiryTest looks to be an intermittent failure with the test itself. I see it failed the exact same way here https://ci.infinispan.org/job/Infinispan/job/master/2346/testReport/junit/org.jsr107.tck.expiry/CacheExpiryTest/expire_whenCreated_TouchedExpiryPolicy_2/

**Comment by pruivo** [comment]
integrated! thanks @wburns !
